### DISCUSSION TITLE
Port rtmfp-cpp to Wasm through WASI

### DISCUSSION THREAD

**Discussion by ROBERT-MCDOWELL** [zenomt/rtmfp-cpp#18] [Ideas]
As browsers don't allow UDP datagramm socket why not to open a discussion at WASI community to allow to create RTMFP and other sockets and allow Wasm to increase its capabilities?

**Comment by zenomt** [comment]
the core of the RTMFP library should be highly portable, since i've been trying to stick to plain C++ 11 and standard socket headers (just for IP address handling). the parts of the library that are really Unix/POSIX and OpenSSL specific are:
* `EPollRunLoop` (just for Linux)
* `SelectRunLoop`
* `Performer`
* `PosixPlatformAdapter` and `PerformerPosixPlatformAdapter`
* `FlashCryptoAdapter_OpenSSL`

last i heard, the core RTMFP library (minus the Unix-specific modules above) compiles on Windows with very few `#ifdef/ifndef _WIN32`s in the code. i haven't heard reports of whether anyone has successfully made a Windows platform adapter though.

i'm not familiar with WASI and have never experimented with Wasm, so i don't know what the portability pitfalls would be. i would expect only minor modifications, if any, would be needed to correct any general portability oversights of the core library. i'm happy to fix all general C/C++11 portability misses, and to adjust the CMakeLists.txt file so that the non-portable modules are only built for Unix-like environments (vs right now where `if(NOT WIN32)` means "is Unix"). however, i really really don't want to end up with a bunch of "`#ifdef WASI`" (or whatever) all through the source code.

the major work, i'd imagine, would be in making WASI platform adapters (to talk to sockets and run loops and stuff) and concrete crypto adapters (assuming OpenSSL isn't available). just like for Windows, i don't want to add WASI-specific modules to this repo, mostly because i don't have the time or expertise to develop, test, and maintain them. i'd be happy to provide links to other folks' repositories that provide these adapters (and any other glue or patches or builders or whatever) though.

**Comment by ROBERT-MCDOWELL** [comment]
Thanks Michael for your very detailed answer. As I know WASI is actively in development, meaning it's a standard that stills accept new idea and feature request to become a standard.  Most of these developers are from google but the community is wide. I suggest to follow and communicate with them if you wish https://github.com/WebAssembly/WASI. the goal of this project is apparently to make wasm be able to request some System Interface and expand its capabilities so it avoids OS differences. It recalls me like Adobe Air in fact.
https://github.com/WebAssembly/WASI
They stilll are open mind to accept new API request, so maybe we can apply it for RMTFP / socket portability. I remember in Flash when p2p multicast rtmfp was involved in the web browser app a confirmation to use p2p was show to the client, so why not make a proposition like to wasm/wasi so it will surely help also the Ruffle project to implement RTMPx/RTMFP in a clearer manner https://github.com/ruffle-rs/ruffle/

**Comment by zenomt** [comment]
it seems like https://github.com/WebAssembly/wasi-sockets is already proposing UDP and TCP sockets, as well as `select` (and other options, like `poll`, `epoll`, `kqueue`), and sounds like they're looking for mostly POSIX compatibility. https://github.com/WebAssembly/wasi-crypto should cover the necessary cryptography (of course, the cryptographic needs of RTMFP for Flash compatibility could also be provided by many other C crypto libraries compiled for Wasm).

**Reply by ROBERT-MCDOWELL** [reply]
yes it  looks promising
